src/rsz/src/SpefWriter.cc

```diff
@@ -0,0 +1,246 @@
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "rsz/SpefWriter.hh"
+
+#include <fstream>
+#include <iostream>
+
+#include "db_sta/dbNetwork.hh"
+#include "sta/Corner.hh"
+#include "sta/Parasitics.hh"
+#include "sta/Units.hh"
+#include "utl/Logger.h"
+
+namespace rsz {
+
+using std::map;
+using utl::RSZ;
+
+SpefWriter::SpefWriter(Logger* logger,
+                       dbSta* sta,
+                       std::map<Corner*, std::ostream*>& spef_files)
+    : logger_(logger),
+      sta_(sta),
+      network_(sta_->getDbNetwork()),
+      parasitics_(sta_->parasitics()),
+      spef_files_(spef_files)
+{
+}
+
+void SpefWriter::writeSpefHeader(Corner* corner)
+{
+  auto it = spef_files_.find(corner);
+  if (it == spef_files_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+  stream << "*SPEF \"ieee 1481-1999\"" << '\n';
+  stream << "*DESIGN \"" << network_->block()->getName() << "\"" << '\n';
+  stream << "*DATE \"11:11:11 Fri 11 11, 1111\"" << '\n';
+  stream << "*VENDOR \"The OpenROAD Project\"" << '\n';
+  stream << "*PROGRAM \"OpenROAD\"" << '\n';
+  stream << "*VERSION \"1.0\"" << '\n';
+  stream << "*DESIGN_FLOW \"NAME_SCOPE LOCAL\" \"PIN_CAP NONE\"" << '\n';
+  stream << "*DIVIDER /" << '\n';
+  stream << "*DELIMITER :" << '\n';
+  stream << "*BUS_DELIMITER []" << '\n';
+
+  auto units = network_->units();
+  std::string time_unit = std::string(units->timeUnit()->suffix());
+  std::string cap_unit = std::string(units->capacitanceUnit()->suffix());
+  std::string res_unit = std::string(units->resistanceUnit()->suffix());
+  std::transform(
+      time_unit.begin(), time_unit.end(), time_unit.begin(), ::toupper);
+  std::transform(cap_unit.begin(), cap_unit.end(), cap_unit.begin(), ::toupper);
+  std::transform(res_unit.begin(), res_unit.end(), res_unit.begin(), ::toupper);
+
+  stream << "*T_UNIT 1 " << time_unit << '\n';
+  stream << "*C_UNIT 1 " << cap_unit << '\n';
+  stream << "*R_UNIT 1 " << res_unit << '\n';
+  stream << '\n';
+}
+
+void SpefWriter::writeSpefPorts(Corner* corner)
+{
+  auto it = spef_files_.find(corner);
+  if (it == spef_files_.end()) {
+    return;
+  }
+  std::ostream& stream = *it->second;
+
+  auto pin_iter = network_->pinIterator(network_->topInstance());
```

Comment:
Using a std::unique_ptr here, as suggested by @maliberty, is clearer and remove the need for the `delete` in the end of the function.